[FEATURE] Filtrer les OIDC providers filtrés par app sur Pix App et Pix Admin (PIX-19809)

lego-technix opened the pull request:
## 🍂 Problème

Le code actuel se base sur les colonnes `enabled` et `enabledForPixAdmin` pour déterminer quels SSO sont à proposer respectivement pour Pix App et Pix Admin. Ce traitement doit gérer des cas spéciaux au lieu d’être générique et ne permet pas de proposer des SSO pour d’autres applications que Pix App et Pix Admin.

Il faut bien prendre en compte la problématique de cache HTTP car la route va maintenant être appelée de manière générique par toutes les applications et à la fois sur le domaine `.fr` et le domaine `.org`.

## 🌰 Proposition

Filtrer les OIDC providers récupérés par Pix App, Pix Admin et autres en fonction des applications (en se basant sur la `requestedApplication`),

### HTTP Cache

Pour garantir qu’un même frontal, Pix App ou Pix Orga, sur des domaines différents, `.fr` et `.org`, ne va pas utiliser la même version en cache de la route `/api/oidc/identity-providers/` on rend cette route appelable avec des URL différents : 
* `/api/oidc/identity-providers/fr`
* `/api/oidc/identity-providers/org`

### Prochaines tâches à faire

Dans une tâche suivante on mettra à jour les enregistrements de la table `oidc-providers` ayant `enabledForPixAdmin=true` et on supprimera la colonne `enabledForPixAdmin`.

Dans une PR suivante on supprimera le code manipulant la propriété `enabledForPixAdmin`.

## 🍁 Remarques

Penser à utiliser l’option `Hide whitespace` pour la relecture des commits qui réorganisent les contextes de tests.

## 🪵 Pour tester

Les tests sont à effectuer en local.

* Aller sur Pix App sur le TLD `.fr`, constater que les SSO proposés sont bien ceux attendus et vérifier que tout fonctionne comme attendu de la connexion à la déconnexion,
* Aller sur Pix App sur le TLD `.org`, constater que les SSO proposés sont bien ceux attendus et vérifier que tout fonctionne comme attendu de la connexion à la déconnexion,
* Aller sur Pix Admin, constater que les SSO proposés sont bien ceux attendus et vérifier que tout fonctionne comme attendu de la connexion à la déconnexion.
* Vérifier que l’appel `GET /api/oidc/identity-providers/` fonctionne toujours et renvoie la même réponse que `GET /api/oidc/identity-providers/org` ou `GET /api/oidc/identity-providers/` de manière à ce que les déconnexions fonctionnent toujours pour Pix Admin et pour les anciennes versions des frontaux Pix App, Pix Orga.

lego-technix committed:
refactor(api): remove a long time unused query param

pix-bot-github commented:
Choisir les applications à déployer :

- [x] [API](https://api-pr14046.review.pix.fr/api/) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr14046/environment) <!-- pix-api-review -->
- [x] [App (.fr)](https://app-pr14046.review.pix.fr) – [App (.org)](https://app-pr14046.review.pix.org) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-app-review-pr14046/environment) <!-- pix-app-review -->
- [x] [Orga (.fr)](https://orga-pr14046.review.pix.fr) – [Orga (.org)](https://orga-pr14046.review.pix.org) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-orga-review-pr14046/environment) <!-- pix-orga-review -->
- [ ] [Certif (.fr)](https://certif-pr14046.review.pix.fr) – [Certif (.org)](https://certif-pr14046.review.pix.org) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-certif-review-pr14046/environment) <!-- pix-certif-review -->
- [ ] [Junior](https://junior-pr14046.review.pix.fr) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-junior-review-pr14046/environment) <!-- pix-junior-review -->
- [x] [Admin](https://admin-pr14046.review.pix.fr) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-admin-review-pr14046/environment) <!-- pix-admin-review -->
- [ ] [API MaDDo](https://pix-api-maddo-review-pr14046.osc-fr1.scalingo.io/api/) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-maddo-review-pr14046/environment) <!-- pix-api-maddo-review -->
- [ ] [Audit Logger](https://pix-audit-logger-review-pr14046.osc-fr1.scalingo.io/api/) | [👩‍💻 Dashboard Scalingo](https://dashboard.scalingo.com/apps/osc-fr1/pix-audit-logger-review-pr14046/environment) <!-- pix-audit-logger-review -->

> [!IMPORTANT]
> N'oubliez pas de déployer l'API pour pouvoir accéder aux fronts et/ou à l’API MaDDo.

lego-technix committed:
feat(api): add applicationTld property to RequestedApplication

lego-technix committed:
feat(api): add application and applicationTld properties to OidcAuthenticationService

lego-technix committed:
feat(api): replace getReadyOidcProviderServices by new generic function getReadyOidcProviderServicesByRequestedApplication

lego-technix committed:
test(api): simplify test contexts

lego-technix committed:
test(api): simplify test contexts

lego-technix committed:
test(api): make createMockedTestOidcProvider configurable wrt application and applicationTld

lego-technix committed:
test(api): remove useless code concerning cookies, it’s done already

lego-technix committed:
test(api): add a test context for "POST /api/oidc/token"

to semantically separate generic code from application specific code. This will
be handy in further code evolution.

lego-technix committed:
feat(api): replace getReadyOidcProviderServicesForPixAdmin by new generic function getReadyOidcProviderServicesByRequestedApplication

bpetetot commented on api/src/identity-access-management/domain/services/oidc-authentication-service-registry.js:
```diff
@@ -36,19 +32,16 @@ export class OidcAuthenticationServiceRegistry {
+  getReadyOidcProviderServicesByRequestedApplication(requestedApplication) {
+    return this.#readyOidcProviderServicesByRequestedApplications[
+      generateGroupByKey(requestedApplication.applicationName, requestedApplication.applicationTld)
+    ];
```
```suggestion
  getReadyOidcProviderServicesByRequestedApplication(requestedApplication) {
    const groupByKey = generateGroupByKey(requestedApplication.applicationName, requestedApplication.applicationTld);
    return this.#readyOidcProviderServicesByRequestedApplications[groupByKey];
  }
```

lego-technix replied:
FAIT : Code réécrit comme suggéré.

bpetetot commented on api/src/identity-access-management/domain/services/oidc-authentication-service-registry.js:
```diff
@@ -36,19 +32,16 @@ export class OidcAuthenticationServiceRegistry {
-    const oidcProviderService = services.find((service) => identityProviderCode === service.code);
+    const oidcProviderService = this.#readyOidcProviderServicesByRequestedApplications[
+      generateGroupByKey(requestedApplication.applicationName, requestedApplication.applicationTld)
+    ].find((service) => identityProviderCode === service.code);
```
On peut réutiliser `this.getReadyOidcProviderServicesByRequestedApplication`

```suggestion
    const appOidcProviderServices = this.getReadyOidcProviderServicesByRequestedApplication(requestedApplication);
    const oidcProviderService = appOidcProviderServices.find((service) => identityProviderCode === service.code);
```

lego-technix replied:
FAIT : Code réécrit comme suggéré.

bpetetot commented on api/src/identity-access-management/domain/services/oidc-authentication-service-registry.js:
```diff
@@ -80,18 +73,23 @@ export class OidcAuthenticationServiceRegistry {
+
+    this.#readyOidcProviderServicesByRequestedApplications = Object.groupBy(
+      this.#allOidcProviderServices.filter(
+        (oidcProviderService) => oidcProviderService.isReady || oidcProviderService.isReadyForPixAdmin,
```
Question: pourquoi on a un cas spécial avec `isReadyForPixAdmin` ? Quelle est la différence entre `isReady` et `isReadyForPixAdmin`

lego-technix replied:
Les identityProviders pour Pix Admin n’ont pas `isReady`, ils ont `isReadyForPixAdmin` (qui dépend de la colonne `enabledForPixAdmin` de `oidc-providers`).

On pourra supprimer la propriété `isReady` lorsque la tâche suivante, qui est ci-dessus dans la description de la solution de cette PR 👆, aura été accomplie : 
> Dans une tâche suivante on mettra à jour les enregistrements de la table oidc-providers ayant enabledForPixAdmin=true et on supprimera la colonne enabledForPixAdmin.

bpetetot commented on api/src/identity-access-management/infrastructure/utils/network.js:
```diff
@@ -82,7 +85,7 @@ export class RequestedApplication {
     if (url.hostname == 'localhost') {
       applicationName = localhostApplicationPortMapping[url.port];
-      return new RequestedApplication(applicationName);
+      return new RequestedApplication({ applicationName });
```
Question: Est-ce qu'il ne faut pas mettre `applicationTld: ''` ici, car quand on calcul les clés dans le registry on aurait `appundefined`, non ?

lego-technix replied:
FAIT : La correction nécessaire de passer `applicationTld: ''` au constructeur de RequestedApplication a été réalisée.

bpetetot commented:
✅ Tests fonctionnels OK

EmmanuelleBonnemay commented:
> ✅ Tests fonctionnels OK

+1

lego-technix committed:
feat(mon-pix): call modified route /api/oidc/identity-providers/{.fr,.org} to not share cache

between application on .fr and application on .org (cache buster).

lego-technix committed:
feat(orga): call modified route /api/oidc/identity-providers/{.fr,.org} to not share cache

between application on .fr and application on .org (cache buster).

lego-technix committed:
feat(api): modify route to have /api/oidc/identity-providers/{.fr,.org} to not share cache

(cache buster)

lego-technix committed:
test(api): make tests independent of other tests

lego-technix committed:
test(orga): fix test title typo

lego-technix committed:
test(mon-pix): fix test title typo

lego-technix committed:
test(admin): fix test title typo

bpetetot reviewed (approved):
✅ Revu et testé

pix-service commented:
:tada: This PR is included in version 5.258.0 :tada:

The release is available on [GitHub release](https://github.com/1024pix/pix/releases/tag/v5.258.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
